core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceParamUtil.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.linecorp.armeria.common.HttpParameters.EMPTY_PARAMETERS;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Ascii;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.common.MediaType;
+
+import io.netty.handler.codec.http.HttpConstants;
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+public final class AnnotatedHttpServiceParamUtil {
+    private static final Logger logger = LoggerFactory.getLogger(AnnotatedHttpServiceParamUtil.class);
+
+    private AnnotatedHttpServiceParamUtil() { }
+
+    /**
+     * Returns a map of parameters decoded from a request.
+     *
+     * <p>Usually one of a query string of a URI or URL-encoded form data is specified in the request.
+     * If both of them exist though, they would be decoded and merged into a parameter map.</p>
+     *
+     * <p>Names and values of the parameters would be decoded as UTF-8 character set.</p>
+     *
+     * @see QueryStringDecoder#QueryStringDecoder(String, boolean)
+     * @see HttpConstants#DEFAULT_CHARSET
+     */
+    public static HttpParameters httpParametersOf(ServiceRequestContext ctx,
+                                                  HttpHeaders reqHeaders,
```

Comment:
is `HttpHeaders` needed while we could get header from `AggregatedHttpMessage`?

Reply:
This code is extracted from `AnnotatedHttpServiceMethod.java`, and in `AnnotatedHttpServiceMethod`, `HttpHeaders` is retrieved from `HttpRequest.headers()` instead of `AggregatedHttpMessage.headers()`.
I am not sure if I can change the original logic of `AnnotatedHttpServiceMethod`